### ISSUE TITLE
Cochran-Armitage trend test

### ISSUE THREAD

**Issue by sbwiecko** [statsmodels/statsmodels#8404]
#### Describe the bug

I am getting different statistical results using the statsmodels test_ordinal_association method compared to the CochranArmitageTest method from DescTools in R.

#### Code Sample, a copy-pastable example if possible

```python
import statsmodels.api as sm

table_tumors = pd.DataFrame(
    data = {
        'control': [a:=0, 50-a],
        'low'    : [b:=1, 50-b],
        'medium' : [c:=3, 50-c],
        'high'   : [d:=6, 50-d]
    },
    index = ['tumor', 'no tumor'],
)

table = sm.stats.Table(table_tumors)

print(table.test_ordinal_association(
    row_scores=np.array([1,0]),
    col_scores=np.array([0,1,2,3]),
))
```

#### Expected Output

Using R, we obtain the following results:

```R
library(DescTools)

dose <- matrix(c(0,1,3,6,50,49,47,44), byrow=TRUE, nrow=2, dimnames=list(resp=c(1,0), dose=0:3))

Desc(dose)
#CochranArmitageTest(dose)
#
#        Cochran-Armitage test for trend
#
#data:  dose
#Z = 2.9019, dim = 4, p-value = 0.003709
#alternative hypothesis: two.sided
```

Using statsmodels, we obtain:

```
null_mean   15.71072319201995
null_sd     3.5390390971784127
pvalue      0.008669937476153359
statistic   25.0
zscore      2.624801973899118
```

#### Output of ``import statsmodels.api as sm; sm.show_versions()``

```
INSTALLED VERSIONS
------------------
Python: 3.10.7.final.0

statsmodels
===========

Installed: 0.13.2 (C:\Users\Sébastien\AppData\Local\Programs\Python\Python310\lib\site-packages\statsmodels)
Required Dependencies
=====================

cython: Not installed
numpy: 1.23.2 (C:\Users\Sébastien\AppData\Local\Programs\Python\Python310\lib\site-packages\numpy)
scipy: 1.9.1 (C:\Users\Sébastien\AppData\Local\Programs\Python\Python310\lib\site-packages\scipy)
pandas: 1.4.4 (C:\Users\Sébastien\AppData\Local\Programs\Python\Python310\lib\site-packages\pandas)
    dateutil: 2.8.2 (C:\Users\Sébastien\AppData\Local\Programs\Python\Python310\lib\site-packages\dateutil)
patsy: 0.5.2 (C:\Users\Sébastien\AppData\Local\Programs\Python\Python310\lib\site-packages\patsy)

Optional Dependencies
=====================

matplotlib: 3.5.3 (C:\Users\Sébastien\AppData\Local\Programs\Python\Python310\lib\site-packages\matplotlib)
    backend: TkAgg
cvxopt: Not installed
joblib: 1.1.0 (C:\Users\Sébastien\AppData\Local\Programs\Python\Python310\lib\site-packages\joblib)

Developer Tools
================

IPython: 8.5.0 (C:\Users\Sébastien\AppData\Local\Programs\Python\Python310\lib\site-packages\IPython)
    jinja2: 3.1.2 (C:\Users\Sébastien\AppData\Local\Programs\Python\Python310\lib\site-packages\jinja2)
sphinx: Not installed
    pygments: 2.13.0 (C:\Users\Sébastien\AppData\Local\Programs\Python\Python310\lib\site-packages\pygments)
pytest: Not installed
virtualenv: Not installed

[paste the output of ``import statsmodels.api as sm; sm.show_versions()`` here below this line]
```
</details>

**Comment by josef-pkt** [comment]
main problem is that contingency tables default to shifting zeros, add 0.5 to all zeros, but not to other entries


from coin package
```
> lbl_test(as.table(dose))

        Asymptotic Linear-by-Linear Association Test

data:  dose (ordered) by resp (1, 2)
Z = 2.8946, p-value = 0.003796
alternative hypothesis: two.sided
```

statsmodels without shift_zeros

```
table = sm.stats.Table(table_tumors, shift_zeros=False)
​
print(table.test_ordinal_association(
    row_scores=np.array([1,0]),
    col_scores=np.array([0,1,2,3]),
))
​
null_mean   15.0
null_sd     3.454659660058306
pvalue      0.0037959220933673713
statistic   25.0
zscore      2.894641146743591
```

however, 2 other R packages give

```
stats

> prop.trend.test(dose[1, ], dose[1, ] + dose[2, ])

        Chi-squared Test for Trend in Proportions

data:  dose[1, ] out of dose[1, ] + dose[2, ] ,
 using scores: 1 2 3 4
X-squared = 8.4211, df = 1, p-value = 0.003709


DescTools

> CochranArmitageTest(dose)

        Cochran-Armitage test for trend

data:  dose
Z = 2.9019, dim = 4, p-value = 0.003709
alternative hypothesis: two.sided
```

**Comment by josef-pkt** [comment]
sorry for the delay. It took me some time to get back to this.

I don't know why R stats and DescTools differ from the linear by linear test

**Comment by sbwiecko** [comment]
> main problem is that contingency tables default to shifting zeros, add 0.5 to all zeros, but not to other entries
> 
> from coin package
> 
> ```
> > lbl_test(as.table(dose))
> 
>         Asymptotic Linear-by-Linear Association Test
> 
> data:  dose (ordered) by resp (1, 2)
> Z = 2.8946, p-value = 0.003796
> alternative hypothesis: two.sided
> ```
> 
> statsmodels without shift_zeros
> 
> ```
> table = sm.stats.Table(table_tumors, shift_zeros=False)
> ​
> print(table.test_ordinal_association(
>     row_scores=np.array([1,0]),
>     col_scores=np.array([0,1,2,3]),
> ))
> ​
> null_mean   15.0
> null_sd     3.454659660058306
> pvalue      0.0037959220933673713
> statistic   25.0
> zscore      2.894641146743591
> ```
> 
> however, 2 other R packages give
> 
> ```
> stats
> 
> > prop.trend.test(dose[1, ], dose[1, ] + dose[2, ])
> 
>         Chi-squared Test for Trend in Proportions
> 
> data:  dose[1, ] out of dose[1, ] + dose[2, ] ,
>  using scores: 1 2 3 4
> X-squared = 8.4211, df = 1, p-value = 0.003709
> 
> 
> DescTools
> 
> > CochranArmitageTest(dose)
> 
>         Cochran-Armitage test for trend
> 
> data:  dose
> Z = 2.9019, dim = 4, p-value = 0.003709
> alternative hypothesis: two.sided
> ```

It does not add 0.5 to all cells but only to zeros:
```python
    shift_zeros : bool
        If True and any cell count is zero, add 0.5 to all values
        in the table.

   (...)

    def __init__(self, table, shift_zeros=True):

        self.table_orig = table
        self.table = np.asarray(table, dtype=np.float64)

        if shift_zeros and (self.table.min() == 0):
            self.table[self.table == 0] = 0.5
```

**Comment by josef-pkt** [comment]
and because it's fun, a combined version proposed by copilot
at least the score test is wrong, other parts look superficially correct

The advantage of figuring out the regression equivalents, we can add other variables as controls.



```
import numpy as np
import statsmodels.api as sm
from scipy import stats


# ------------------------------------------------------------
# Classical Cochran–Armitage
# ------------------------------------------------------------
def cochran_armitage_classic(successes, totals, scores=None, two_sided=True):
    """
    Classical Cochran–Armitage trend test for a 2×k table.
    """
    x = np.asarray(successes, dtype=float)
    n = np.asarray(totals, dtype=float)
    if scores is None:
        t = np.arange(1, len(x) + 1, dtype=float)
    else:
        t = np.asarray(scores, dtype=float)

    N = n.sum()
    p_hat = x.sum() / N
    t_bar = (n * t).sum() / N

    num = (t * (x - n * p_hat)).sum()
    denom = np.sqrt(p_hat * (1 - p_hat) * (n * (t - t_bar)**2).sum())

    z = num / denom
    chi2 = z**2

    if two_sided:
        p_value = 2 * (1 - stats.norm.cdf(abs(z)))
    else:
        p_value = 1 - stats.norm.cdf(z)

    return {"z": z, "chi2": chi2, "p_value": p_value}


# ------------------------------------------------------------
# GLM score test (true Cochran–Armitage)
# ------------------------------------------------------------
def cochran_armitage_glm_score(successes, totals, scores=None, two_sided=True):
    x = np.asarray(successes, dtype=float)
    n = np.asarray(totals, dtype=float)
    if scores is None:
        t = np.arange(1, len(x) + 1, dtype=float)
    else:
        t = np.asarray(scores, dtype=float)

    endog = np.column_stack([x, n - x])
    exog = sm.add_constant(t)

    model = sm.GLM(endog, exog, family=sm.families.Binomial())
    res = model.fit(disp=0)

    st = res.score_test(term=1)
    chi2 = float(st.statistic)
    z_signed = np.sign(res.params[1]) * np.sqrt(chi2)

    if two_sided:
        p_value = float(st.pvalue)
    else:
        p_value = 1 - stats.norm.cdf(z_signed)

    return {"z": z_signed, "chi2": chi2, "p_value": p_value}


# ------------------------------------------------------------
# GLM Wald test
# ------------------------------------------------------------
def trend_test_glm_wald(successes, totals, scores=None, two_sided=True):
    x = np.asarray(successes, dtype=float)
    n = np.asarray(totals, dtype=float)
    if scores is None:
        t = np.arange(1, len(x) + 1, dtype=float)
    else:
        t = np.asarray(scores, dtype=float)

    endog = np.column_stack([x, n - x])
    exog = sm.add_constant(t)

    model = sm.GLM(endog, exog, family=sm.families.Binomial())
    res = model.fit(disp=0)

    beta = res.params[1]
    se = res.bse[1]
    z = beta / se
    chi2 = z**2

    if two_sided:
        p_value = 2 * (1 - stats.norm.cdf(abs(z)))
    else:
        p_value = 1 - stats.norm.cdf(z)

    return {"z": z, "chi2": chi2, "p_value": p_value}


# ------------------------------------------------------------
# GLM Likelihood‑ratio test
# ------------------------------------------------------------
def trend_test_glm_lrt(successes, totals, scores=None):
    x = np.asarray(successes, dtype=float)
    n = np.asarray(totals, dtype=float)
    if scores is None:
        t = np.arange(1, len(x) + 1, dtype=float)
    else:
        t = np.asarray(scores, dtype=float)

    endog = np.column_stack([x, n - x])

    # Null model: intercept only
    exog_null = np.ones((len(x), 1))
    model_null = sm.GLM(endog, exog_null, family=sm.families.Binomial())
    res_null = model_null.fit(disp=0)

    # Full model: intercept + score
    exog_full = sm.add_constant(t)
    model_full = sm.GLM(endog, exog_full, family=sm.families.Binomial())
    res_full = model_full.fit(disp=0)

    lrt = 2 * (res_full.llf - res_null.llf)
    p_value = 1 - stats.chi2.cdf(lrt, df=1)

    return {"chi2": lrt, "p_value": p_value}


# ------------------------------------------------------------
# Combined harness
# ------------------------------------------------------------
def run_all_trend_tests(successes, totals, scores=None):
    return {
        "classic_CA": cochran_armitage_classic(successes, totals, scores),
        "glm_score": cochran_armitage_glm_score(successes, totals, scores),
        "glm_wald": trend_test_glm_wald(successes, totals, scores),
        "glm_lrt": trend_test_glm_lrt(successes, totals, scores),
    }


# ------------------------------------------------------------
# Example
# ------------------------------------------------------------
if __name__ == "__main__":
    n = np.array([50, 60, 55, 65])
    x = np.array([5, 12, 18, 30])

    results = run_all_trend_tests(x, n)
    for name, res in results.items():
        print(f"\n{name}:")
        print(res)
```

**Comment by josef-pkt** [comment]
AI spam
I should have started a new issue instead of using a FAQ issue.

to finish this part (other versions will follow in new issues or prs):

The following is a, not yet fully cleaned up, version for the classical Cochran-Armitage trend test.
Advantage: It's vectorize for observations in rows and different series in columns.
(This assumes as in other vectorized hypothesis tests that each series has the same length, which is in this case the number of ordinal levels.)
Trend tests that are based on models, GLM, WLS, cannot be automatically vectorized.

successes, totals need to have the same shape or the same number of dimensions (1 or 2) and broadcast.
scores can be None, 1-dim or 2-dim. (If scores is 2dim, then levels need to be in rows.)
 

```
from statsmodels.stats.base import HolderTuple
from statsmodels.stats.weightstats import _zstat_generic2


def test_cochran_armitage_trend(successes, totals, scores=None, alternative="two-sided"):
    x = np.asarray(successes, dtype=float)
    n = np.asarray(totals, dtype=float)
    if scores is None:
        t = np.arange(1, x.shape[0] + 1, dtype=float)
    else:
        t = np.asarray(scores, dtype=float)

    if x.ndim > 1 and t.ndim == 1:
        t = t[:, None]

    nobs_total = n.sum(0)
    p_hat = x.sum(0) / nobs_total

    t_bar = (n * t).sum(0) / nobs_total
    num = (t * (x - n * p_hat)).sum(0)
    denom = np.sqrt(p_hat * (1 - p_hat) * (n * (t - t_bar) ** 2).sum(0))

    z = num / denom
    chi2 = z**2

    # allow different terminology and convert
    if alternative == "increasing":
        alternative = "larger"
    elif alternative == "decreasing":
        alternative = "smaller"

    _, p_value = _zstat_generic2(z, 1, alternative=alternative)

    return HolderTuple(
        statistic=z, 
        chi2=chi2,
        pvalue=p_value,
        distr="norm"
        )
```

**Comment by josef-pkt** [comment]
one more: Cochran-Armitage trend test as score test in GLM Binomial
This produces the same results (pvalue) as CA classic standalone test. 
So, CAT tests really is the score_test in Logit/GLM-Binomial.

aside based on a quick try: 
I get the same score test pvalue if I use Probit link instead of default canonical, logit link. Why?
It might be related to the property that score_test is robust to reparameterization, but I don't see the reason with nonlinear link.)
If we allow for additional regressors (confounders), then providing an option for the link would be useful. It would still be a score_test for trend but in a different model, not anymore exactly the same as CAT test. 

note: score_test returns tuple and not a results class #9785

```
def cochran_armitage_glm_score(successes, totals, scores=None):
    x = np.asarray(successes, float)
    n = np.asarray(totals, float)
    if scores is None:
        t = np.arange(1, len(x) + 1, dtype=float)
    else:
        t = np.asarray(scores, float)

    endog = np.column_stack([x, n - x])
    exog = np.ones(len(endog))

    model = GLM(endog, exog, family=families.Binomial())
    res = model.fit(disp=0)

    st = res.score_test(exog_extra=t)
    chi2, pvalue, df = st
    #chi2 = float(st.statistic)
    Z = np.nan #np.sign(res.params[1]) * np.sqrt(chi2)
    return {"Z": Z, "chi2": chi2, "p_value": pvalue}
```
